streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -565,13 +581,44 @@ private void runLoop() {
 
                 boolean longPoll = totalNumBuffered == 0;
 
-                ConsumerRecords<byte[], byte[]> records = consumer.poll(longPoll ? this.pollTimeMs : 0);
+                ConsumerRecords<byte[], byte[]> records = null;
+
+                try {
+                    records = consumer.poll(longPoll ? this.pollTimeMs : 0);
+                } catch (NoOffsetForPartitionException ex) {
+                    TopicPartition partition = ex.partition();
+                    if (builder.earliestResetTopicsPattern().matcher(partition.topic()).matches()) {
+                        log.info(String.format("stream-thread [%s] setting topic to consume from earliest offset %s", this.getName(), partition.topic()));
+                        consumer.seekToBeginning(ex.partitions());
+                    } else if (builder.latestResetTopicsPattern().matcher(partition.topic()).matches()) {
+                        consumer.seekToEnd(ex.partitions());
+                        log.info(String.format("stream-thread [%s] setting topic to consume from latest offset %s", this.getName(), partition.topic()));
+                    } else {
+
+                        if (originalReset == null || (!originalReset.equals("earliest") && !originalReset.equals("latest"))) {
+                            setState(State.PENDING_SHUTDOWN);
```

Comment:
Fair enough, let me propose an alternative.  

When throwing the `StreamsException`  (wrapping the `NoOffsetForPartitionException`), the `close()` method gets invoked which eventually calls the `setState(State.NOT_RUNNING)` which in turn throws an `IllegalStateException` because we are trying to transition from `RUNNING` to `NOT_RUNNING`.

What if we were to set an instance variable `Throwable underlyingException` and if it is not null,  pass the `underlyingException` variable to the `IllegalStateException' constructor?

That way we can communicate the exact source of the error back to the user.

What do you think?